### PR TITLE
Bl 4587original credits update

### PR DESCRIPTION
**Pull request by JohnThomson**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/1689)

### CODE REVIEW

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 848 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwOTwmNDhqJicdS-mF:-KkwOTwnbGf4AKYWvhiJ:b-17ebrk) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomExe/Book/Book.cs#L848)):*
> ```Smalltalk
> 			{
> 				// Original version of this code that suffers BL_3166
> 				BringBookUpToDateUnprotected(bookDOM, progress);
> ```

This change isn't related except that this method has two long blocks of identical code and I decided it was time to factor out a method.

---

*[src/BloomExe/Book/Book.cs, line 868 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwOgmBLdPXbpPBEC-o:-KkwOgmBLdPXbpPBEC-p:bap2g4a) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomExe/Book/Book.cs#L868)):*
> ```Smalltalk
> 			// Before we update the xmatter, or at least before SynchronizeDataItemsThroughoutDOM(),
> 			// so the updated (to current language settings) version of original copyright notice gets injected.
> 			BookCopyrightAndLicense.UpdateOriginalCopyrightAndLicense(bookDOM, _bookData, _collectionSettings);
> ```

This line and its comments are the only real change to the original BringBookUpToDate()

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 390 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwP2Qogw_9C5I6Ms-t:-KkwP2Qogw_9C5I6Ms-u:b-ujo88b) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomExe/Book/BookCopyrightAndLicense.cs#L390)):*
> ```Smalltalk
> 					preferredLanguageIds, out idOfLanguageUsed);
> 				copyrightNotice = String.Format(originalCopyrightSentence, metadata.CopyrightNotice.Trim()) + " " +
> 				                  originalLicenseSentence;
> ```

Just an auto-format re-wrap

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 515 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwPBwC36whs0O7IjFM:-KkwPBwC36whs0O7IjFN:b9c0str) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d0cb5f7f854d0fefcea2261b1483cfeb7582c825/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L515)):*
> ```Smalltalk
> 					  <div data-book='bookTitle' lang='en'>A really really empty book</div>
> 					</div>");
> 			AssertThatXmlIn.Dom(dom.RawDom).HasNoMatchForXpath("//div[@id='bloomDataDiv' and @data-book='originalCopyrightAndLicense']");
> ```

This test was doubly flawed: the xpath is wrong, so it was not actually checking that we don't get original credits in this case; and in fact we (intentionally) DO get some original credits. So I updated the test in that direction. It doesn't reflect a change in the method tested.

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 518 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwTNSxpMt7HWyDG1_V:-KkwTNSxpMt7HWyDG1_W:bx7iiq9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L518)):*
> ```Smalltalk
> 						</div>
> 					</div>");
> 			AssertThatXmlIn.Dom(dom.RawDom).HasNoMatchForXpath("//div[@id='bloomDataDiv']/div[@data-book='copyright']");
> ```

Here I'm just fixing a bug in the original test.

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 557 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwT-r-0ZLmjsTS02Zr:-KkwT-r-0ZLmjsTS02Zs:bp8zly7) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L557)):*
> ```Smalltalk
> 					  <div data-book='bookTitle' lang='en'>A really really empty book</div>
> 					</div>");
> 			Assert.AreEqual("Adapted from original without a copyright notice. Licensed under CC-BY 4.0.", GetEnglishOriginalCopyrightAndLicense(dom));
> ```

Should we really get a license in the original credits, if the source book somehow didn't have one?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by JohnThomson** [comment]
Don't merge this yet, want to change so the derived data is not stored at all.

File: src/BloomExe/Book/BookCopyrightAndLicense.cs [hunk 1]
```diff
@@ -303,19 +313,50 @@ public static void SetOriginalCopyrightAndLicense(HtmlDom dom, BookData bookData
...
+		/// <summary>
+		/// This is used each time we update xmatter. It assumes originalLicenseUrl, originalCopyright, and originalLicenseNotes have
+		/// already been set while creating the book. It updates the actual displayed originalCopyrightAndLicense to the appropriate
+		/// derived value based on the current language settings of the book. It should be called before we update the DOM from
+		/// the datadiv.
+		/// </summary>
+		/// <param name="dom"></param>
+		/// <param name="bookData"></param>
+		/// <param name="collectionSettings"></param>
+		public static void UpdateOriginalCopyrightAndLicense(HtmlDom dom, BookData bookData, CollectionSettings collectionSettings)
```
Comment (hatton) [inline src/BloomExe/Book/BookCopyrightAndLicense.cs]: I think this should be redone as a method parallel to BookData.SetUpDisplayOfTopicInBook(). It should not write anything to the data-div. Instead, like SetUpDisplayOfTopicInBook(), it should search the book for @data-derived='original....notice' (whatever the exact name is).  

In this way, we make it obvious both in code and in the book itself that these are not new pieces of "data"; they are projections of the original metadata which PR now saves, onto the page.

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 333 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwNcTfhpqUEa3HZSQG-r1-333:-Kl4LXpi0qXc2g8rjG-s:b-j0mzas) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomExe/Book/BookCopyrightAndLicense.cs#L333)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I think this should be redone as a method parallel to BookData.SetUpDisplayOfTopicInBook(). It should not write anything to the data-div. Instead, like SetUpDisplayOfTopicInBook(), it should search the book for @data-derived='original....notice' (whatever the exact name is).  

In this way, we make it obvious both in code and in the book itself that these are not new pieces of "data"; they are projections of the original metadata which PR now saves, onto the page.
</blockquote></details>

I think I've done what you wanted...actually turned out to simplify things, on the whole. The new method is more generic (CopyStringToFieldsInPages)...maybe it could be in HtmlDom? But only used in one place here, and rather parallel to CopyItemToFieldsInPages, though I could not find a clean way to factor out the common logic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by hatton** [comment]
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/BloomExe/Book/Book.cs, line 848 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwOTwmNDhqJicdS-mF:-Kl0N4gojyY_NKhE4gDh:b9y27z0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomExe/Book/Book.cs#L848)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This change isn't related except that this method has two long blocks of identical code and I decided it was time to factor out a method.
</blockquote></details>

great, thanks!

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 199 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KlPCGLSnel-LUOW8jAE:-KlPCGLTLIVBF92kMI3T:b6ls4lv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/41a83cd023c066743f6e7f2e88c01ecb04732376/src/BloomExe/Book/BookCopyrightAndLicense.cs#L199)):*
> ```Smalltalk
> 		private static void CopyStringToFieldsInPages(HtmlDom dom, string key, string val, string lang)
> 		{
> 			var target = dom.SelectSingleNode("//*[@data-derived='" + key + "']");
> ```

The method name has plural "fields", and I think that's right. So this would need to work if the derived string appeared more than once in the document, rather than assuming a "single node". I note that my older CopyItemsToFieldsInPages() has the same problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 199 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KlPCGLSnel-LUOW8jAE:-KltlvEO1l9VSWQhBidX:b724gvl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/41a83cd023c066743f6e7f2e88c01ecb04732376/src/BloomExe/Book/BookCopyrightAndLicense.cs#L199)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

The method name has plural "fields", and I think that's right. So this would need to work if the derived string appeared more than once in the document, rather than assuming a "single node". I note that my older CopyItemsToFieldsInPages() has the same problem.
</blockquote></details>

Done. (Also fixed the other method.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by hatton** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 557 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwT-r-0ZLmjsTS02Zr:-KlvcH0EoEv_pNevBL7V:bquov78) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L557)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Should we really get a license in the original credits, if the source book somehow didn't have one?
</blockquote></details>

I feel I'm being dense (let's blame jet lag) but can you point out to me where the original credits is getting a license when the original book didn't actually have one? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 557 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-KkwT-r-0ZLmjsTS02Zr:-Klxc9b53usTIaBupLcQ:b-bh6huh) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/6ff59a5e68d811431bc61527eb315b5491a8be1e/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L557)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I feel I'm being dense (let's blame jet lag) but can you point out to me where the original credits is getting a license when the original book didn't actually have one? 
</blockquote></details>

Well, the thing we are validating as the value of "originalLicenseAndCredits" includes "Licensed under CC-BY 4.0." although the data-div we started with (which we imagine to have been copied from the shell book) has no div with data-book='licenseUrl' (or any other license info).
Of course that's what Bloom fills in as the license for any book that doesn't already have one when you open it...which is why this test is kind of moot, because I don't see how the shell book can fail to have a licenseUrl...so it's not completely unreasonable to do this, but it seemed odd enough to deserve a little thought.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by andrew-polk** [comment]
Reviewed 1 of 3 files at r1, 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 203 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-Km1Rn7ee3DSio8OFd3J:-Km1Rn7fxX8CO_Zbs6jX:b-6qfbls) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/bfb2e0041cab4920ea89c661081fa7adde2f8e8a/src/BloomExe/Book/BookCopyrightAndLicense.cs#L203)):*
> ```Smalltalk
> 			{
> 				if (target == null)
> 					return;
> ```

Is it possible for one of the results to be null but others not to be? If yes, we should continue here, not return.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 397 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-Km1SQw8Btc5jnYMMdZY:-Km1SQw8Btc5jnYMMdZZ:brugfch) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/bfb2e0041cab4920ea89c661081fa7adde2f8e8a/src/BloomExe/Book/BookCopyrightAndLicense.cs#L397)):*
> ```Smalltalk
> 			}
> 
> 			// The copyright string has to be encoded because it's fed eventually into the XmlNode.InnerXml method, and some people
> ```

Does this comment warrant an update now?

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 393 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-Km1TR63HOsedjVkTtGn:-Km1TR63HOsedjVkTtGo:b-qgejbu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/bfb2e0041cab4920ea89c661081fa7adde2f8e8a/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L393)):*
> enought

enough
Obviously, don't bother if there are no other changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689)*

**Comment by JohnThomson** [comment]
Good catches, thanks.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 203 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-Km1Rn7ee3DSio8OFd3J:-Km26kV0dkVAH6zA3ET-:bipeqad) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/bfb2e0041cab4920ea89c661081fa7adde2f8e8a/src/BloomExe/Book/BookCopyrightAndLicense.cs#L203)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is it possible for one of the results to be null but others not to be? If yes, we should continue here, not return.
</blockquote></details>

Done. (I don't actually think any of them can be, since an xpath searching by attribute can't find anything but elements, but I'm leaving it in because somewhere in my testing it seemed to retrieve a text element.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 397 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-Km1SQw8Btc5jnYMMdZY:-Km27iiir7wi-tZLVXc9:bk8dxa8) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/bfb2e0041cab4920ea89c661081fa7adde2f8e8a/src/BloomExe/Book/BookCopyrightAndLicense.cs#L397)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Does this comment warrant an update now?
</blockquote></details>

Done. (Just removed it, since it's no longer true...the final destination is now InnerText, whose setter handles encoding.

---

*[src/BloomTests/Book/BookCopyrightAndLicenseTests.cs, line 393 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-Km1TR63HOsedjVkTtGn:-Km28CWYdh0NnPD0U03a:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/bfb2e0041cab4920ea89c661081fa7adde2f8e8a/src/BloomTests/Book/BookCopyrightAndLicenseTests.cs#L393)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> enought

enough
Obviously, don't bother if there are no other changes.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-:-Km28EJiC7TsLX1LRZNb:bdl09hi)*

**Comment by andrew-polk** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1689#-:-Km29fCVBu7BPexHu8nn:bnfp4nl)*

**Pull request merged**